Async ECS Access Primitive

MalekiRe opened the pull request:
# Objective

Bevy is missing ergonomic async <-> ecs interactions, especially and specifically access to the ECS from async tasks.

## Solution

This PR adds underlying infrastructure for async runtime agnostic async ecs access through the addition of a new concept called a `bridge`. This allows you to `bridge` an async task with the ecs at sync points which you define and add.
## Testing

A new example `async_bridge_primitive` is added to the async examples modifying the example from the tasks area.

Run that example and read the code to see it in action.

We also have doc-tests for this.

I tested it on x86-linux.

This does *not* work on the web, web support can be added in a follow-up PR in two manners, one which ignores SyncPoints, and the other which doesn't work with `&` and `&mut` variables from the local context. 
## Showcase

```rust
use bevy_app::prelude::*;
use bevy_async::prelude::*;
use bevy_ecs::prelude::*;
use bevy_tasks::AsyncComputeTaskPool;
use bevy_platform::sync::atomic::AtomicBool;
use bevy_platform::sync::atomic::Ordering;
use bevy_platform::sync::Arc;
use bevy_app::ScheduleRunnerPlugin;

struct MySyncPoint;
static ACCESS_RAN: AtomicBool = AtomicBool::new(false);

fn main() {
    let mut app = App::new();
    app.add_plugins((
        AsyncPlugin::default(),
        ScheduleRunnerPlugin::default(),
        TaskPoolPlugin::default(),
    ));

    app.add_systems(Update, async_world_sync_point::<MySyncPoint>);

    app.add_systems(Startup, move |world: Res<AsyncWorld>| {
        let world = world.clone();
        AsyncComputeTaskPool::get()
            .spawn(async move {
                let system_state = world.system_state::<Commands>();
                system_state
                    .bridge(MySyncPoint, |mut commands: Commands| {
                        commands.spawn_empty();
                        ACCESS_RAN.store(true, Ordering::Relaxed);
                    })
                    .await
                    .unwrap();
            })
            .detach();
    });

    app.update();

    assert!(ACCESS_RAN.load(Ordering::Relaxed));
}
```

PROMETHIA-27 commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,432 @@
+    /// HashMap<K, Arc<ConcurrentQueue<V>>> behind a single RwLock.
+    /// - Writers only contend when creating a new key or GC'ing.
+    /// - `push` is non-blocking (unbounded queue).
+    pub struct KeyedQueues<K, V> {
```
Can you expand the doc comment to explain what a KeyedQueue is and why you would use it? And maybe move the info relevant to `inner` onto `inner`.

PROMETHIA-27 commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,432 @@
+);
+
+#[derive(bevy_ecs_macros::Resource, Default)]
+pub(crate) struct SystemParamAppliers(HashMap<TypeId, fn(&mut World)>);
```
Does this permit duplicate systemparams?

PROMETHIA-27 commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,432 @@
+                self.slot.write().unwrap().take();
+            }
+        }
+        unsafe {
```
This unsafe block is too large; shrink it to only the relevant operations (afaictjust the transmute?)

MalekiRe replied:
Hmm I am *not* confident that's how it's done in the bevy engine, that is the broader way that i've seen it being done, but bevy might do it differently, i'll wait for someone to chime in who knows and if not i'll defer to you.

PROMETHIA-27 commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,432 @@
+            // Lifetimes are not used in any actual code optimization, so turning it into a static does not violate any of rust's rules
+            // As *LONG* as we keep it within it's lifetime, which we do here, manually, with our `ClearOnDrop` struct.
+            world_container.write().unwrap().replace((
+                std::mem::transmute(world.as_unsafe_world_cell()),
```
It's good practice to directly annotate the transmute with types, since it's such a powerful operation and it should be both clear at a glance what it is doing and enforced by the compiler rather than allowing type inference to mess anything up.

PROMETHIA-27 commented on examples/ecs/async_ecs.rs:
```diff
@@ -0,0 +1,144 @@
+                // Simulate a delay before task completion
+                println!("delaying for {:?}", delay);
+                Delay::new(delay).await;
+                if let Err(e) = async_access::<
```
Is it possible to rewrite this API to not require the wildcard generic parameters? Or are those unavoidable? Or maybe just set a standard for adding these params at the `EcsTask::new` site? Or annotating the closure's parameter types instead of the generic?

MalekiRe commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,432 @@
+
+impl AsyncTaskId {
+    /// Create a new, unique [`AsyncTaskId`]. Returns [`None`] if the supply of unique
+    /// IDs has been exhausted.
```
This is how WorldId is currently implemented, however it's probably likely that you aren't intended to create a ton of worlds.

Currently if you created one async task every nanosecond it would take 584,542 years to overflow ( if I am doing my calculations correctly ). 

That being said, we could do lots of things to make this number completely unreachable; if we store 4 atomic u64s and combine them then if we ticked them up once every nanosecond it would take 10^137 years to overflow.

PROMETHIA-27 replied:
> 584,542 years to overflow

That's assuming 64 bits; being a usize, on 32-bit machines (which afaik bevy does support, and even supports whatever the gameboy is; might even be 16) that drops to 4 seconds. 1 task per nanosecond might be a pretty unreasonable bar anyway though; at one per frame it's 51 days I think. I'm not sure how many of these tasks you might expect to see in a worst case scenario.

I would maybe just change it to an AtomicU64 to make sure this assumption doesn't get violated on lower bitness hardware?

MalekiRe commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,432 @@
+                };
+                let out;
+                // SAFETY: This is safe because we have a mutex around our world cell, so only one thing can have access to it at a time.
+                #[expect(unused_unsafe)]
```
No it's just more explicit, i could just use a comment to draw attention to it instead though, that probably makes more sense.

cBournhonesque commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,441 @@
+                .contains_key(&task_id)
+            {
+                let system_state = SystemState::<P>::new(world);
+                let cq = ConcurrentQueue::bounded(1);
```
Do we need concurrent queues here?
We remove and push the SystemState always inside this task, right? so could it just be an Option?

MalekiRe replied:
Options are not shareable and mutable across threads so it would have to be an arc<mutex<option>> which i think is less clear than a concurrent queue with size 1, thought both would be non-blocking. So if you do think that it is more clear to use an `Arc<Mutex<Option<T>>>` over a `ConcurrentQueue<T>` i am willing to modify the PR.

MalekiRe commented on crates/bevy_ecs/src/schedule/executor/async.rs:
```diff
@@ -0,0 +1,441 @@
+);
+
+impl WorldAccessRegistry {
+    pub(crate) fn set(&self, world: &mut World, func: impl FnOnce()) -> Option<()> {
```
Updated this with a comment, we do it this way so one cannot forget to `unset` the WorldAccessRegistry. Important for maintainability the less things you have to manually keep track of the better.

PROMETHIA-27 reviewed (approved):
LGTM!

MalekiRe commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -535,7 +535,11 @@ impl Schedule {
         let error_handler = world.default_error_handler();
-
+        #[cfg(feature = "std")]
+        while async_ecs::GLOBAL_WAKE_REGISTRY
```
> I disagree. Resolving this now is important because it is a core philosophical and architectural piece.

Fair enough:

@cart I misunderstood what you meant, I thought you wanted us to schedule async accesses between systems within the async access, in some manner like `async_access(before(some_system), in(Update))` or some such.

What you're proposing is 100% achievable with schedules as is.

Adding a derive sync point could still be useful to try to tighten it up. Imagine this for a moment.
```rust
#[derive(SyncPoint)]
struct AsyncUi;
``` ->
```rust
impl Schedule for AsyncUi {}
impl System for AsyncUi {
  fn run(world: &mut World) {
      world.run_schedule(AsyncUi);
  }
}
```
Then you can achieve everything stated here with even less code.

```rust
#[derive(SyncPoint)]
struct AsyncUi;

// in main / app setup
app.add_systems(PostUpdate, AsyncUi.before(some_system))


// later
pool.spawn(async move {
    let value = async_access::<
        Query<&Node>,
        _,
        _,
    >(
        task_id,
        AsyncUi,
        |nodes: Query<&Node>| {
            // do stuff with nodes
        },
    ).await;
})
```
I *still* think and hold that arbitrarily parallelizing *within* a schedule is a good idea. This is how systems currently work when you add them unless you explicitly order them! It is very useful and I want to eventually support it even if we never add support for explicit ordering on async accesses itself.

We *could* add explicit ordering around individual async accesses, but i think that is out of scope for this first PR. If you still disagree then we can talk about it but we appear to be on the same page about the level of granularity we are talking about.

> 
1. Gets all of the benefits of being a system: supports ordering out of the box, can be labeled / named, will show up in schedule visualizers / editor tooling as a system, plays nicely with system-stepping, will get "systems as entities" features when they come, etc.
2. Requires no changes to schedules / in a way exists "in userspace". This approach also means there can be multiple async approaches, if we so desire, including in third party plugins.
3. Is "zero cost" for developers / apps not using async.

As a response to 1: as shown here we can make a schedule a system that runs the schedule, and make the ergonomics even better than your proposed example.

2. I don't think that this is a good reason to not do something at the scheduler level if and when it can provide superior ergonomics or performance, I do think that eventually supporting parallelization would be a good thing, for intra-task and inter-task-system parallel access. 

3. Currently we do a single `.is_empty` check on a keyed concurrent queue, this is not benchmarked but is going to be pretty close to negligible. 

As a compromise, given that we *currently* are not doing this inter-task parallelism, but with the option to do it in the future ( not exposing an api to users that prevents us from upgrading to it ), how does this approach sound?

```rust
#[derive(SyncPoint)]
struct AsyncUi;
``` ->
```rust
impl Schedule for AsyncUi {}
impl System for AsyncUi {
  fn run(world: &mut World) {
      while ECS_WAKER_LIST.wait().is_some() {}
      world.run_schedule(AsyncUi);
  }
}
```

chescock replied:
> To be more clear, the reason that I don't think `AsyncUi` should _just_ be a system, is that this way allows us to add systems that can run in the same schedule as `AsyncUi` that do not care if they run after, before, or in-between any async_accesses that happen in `AsyncUi`. This is useful for a parallelized implementation over ecs accesses. If `AsyncUi` is a system, then it itself cannot have other systems running inside it that can be ordered arbitrarily around or in parallel with async accesses.

Wouldn't the best solution to that be to support using non-exclusive systems as sync points?  

Using only exclusive systems seems like a good way to reduce scope for the first version, but we may want to generalize it in the future to allow each `async_access` call site to be scheduled independently as a non-exclusive system.  Then async tasks would be able to run concurrently with other systems as long as the parameters don't conflict!

Diddykonga replied:
I do tend to agree with cart in that if we can allow it to be scheduled between systems in a schedule, it would be nice.

I am not sure that the mixed System/Schedule thing should exist, on one hand its nice semantically because it represents the multiple 'systems' that could be polled/run during that sync point, but they don't actually exist as systems in that schedule and you can also add systems to that schedule which is also odd, and is trying to gain the benefits of actual async systems without actually doing it.

Should probably just be what cart suggested, an function that outputs a system that polls the given 'sync point'.

PixelDust22 commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -535,7 +535,7 @@ impl Schedule {
 
         let error_handler = world.default_error_handler();
-
+        while let Some(()) = r#async::GLOBAL_WAKE_REGISTRY.wait(self.label, world) {}
```
So as an user if I want to access a resource `MyCuteResource` from an async context, what I could do is:



```rs
pub struct MyCuteResource(Arc<async_lock::Mutex<MyCuteResourceInner>>;

...
// In sync context
pub fn my_system(resource: Res<MyCuteResource>) {
    let resource = resource.lock_blocking();
    // do some stuff with resource
}

// In async context:
pub fn spawn_tasks(resource: Res<MyCuteResource>) {
    let pool = AsyncComputeTaskPool::get();
    let resource = resource.clone();
    pool.spawn(async move {
        resource.lock().await;
        // do some stuff with resource
    });
}

PixelDust22 replied:
So if I'm understanding this PR correctly:

- Maintaining a list of global accesses
- block conflicting systems from running until global accesses are released

Is basically equivalent to building your own mutex over all resources. So I guess my question is: How is this better than just allowing the user to wrap the resource that they need accessing inside their own mutexes? Or have the bevy engine explicitly wrap all resources in `async_lock::Mutex`?

alice-i-cecile commented on crates/bevy_ecs/examples/simple_async_ecs.rs:
```diff
@@ -0,0 +1,47 @@
+//! In this example we will simulate a population of entities. In every tick we will:
```
TBH, I don't think this example should exist at all. Subcrate examples are extremely hard to discover!

alice-i-cecile commented on examples/ecs/async_ecs.rs:
```diff
@@ -0,0 +1,176 @@
+
+/// Resource holding the mesh handle for the box (used for spawning cubes)
+#[derive(Resource, Deref)]
+struct BoxMeshHandle(Handle<Mesh>);
```
Nit: we can combine these into a single resource with two fields.

MalekiRe committed:
redid everything, brand new pr effectively.

dlom committed:
Rename everything. Still need a docs/organizational pass

`AsyncBridge` -> `AsyncWorld` (with new private `StrongAsyncWorld`)

`AsyncBridge::new` -> `AsyncWorld::system_state`

`EcsAccess` -> `AsyncSystemState`

`EcsAccess::run` -> `AsyncSystemState::bridge`

`drive_async_bridge -> async_world_sync_point`

MalekiRe committed:
Merge pull request #2 from dlom/mold/async-ecs-again

Rename everything. Still need a docs/organizational pass

MalekiRe committed:
reorganized some types including wait_barrier and added more documentation and fixed cit

MalekiRe committed:
fix ci

MalekiRe committed:
change the example to not compile on wasm ( for now )

MalekiRe commented on examples/ecs/async_ecs.rs:
```diff
@@ -0,0 +1,176 @@
+//! A minimal example showing how to perform asynchronous work in Bevy
```
Let me know if the new example is minimal enough

MalekiRe committed:
fix no_std support to make sure it compiles.

MalekiRe committed:
added a unit test for no_std

mrchantey reviewed (approved):
The inner workings are above my pay grade but for an initial release I'm very happy with the public api. Looks like it exposes everything my async libraries need for integration.

Will-Carson reviewed (approved):
This seems pretty straightforward. The overall surface area is pretty small. I look forward to stress testing it.

alice-i-cecile commented on examples/async_tasks/async_bridge_primitive.rs:
```diff
@@ -0,0 +1,117 @@
+//! asynchronously.
+//!
+//! Unlike the channel-based approach (where tasks send results directly via a communication
+//! channel) or the direct approach in `async_compute`, this example uses the ecs <-> async bridge.
```
Nit: clarify that `async_compute` is an example.

andriyDev replied:
Done.

alice-i-cecile commented on crates/bevy_async/src/bridge_future.rs:
```diff
@@ -0,0 +1,233 @@
+                            system_state: self.system_state.clone(),
+                        },
+                    )
+                    .ok()
```
Cleanup: do not call .ok().unwrap()! This just destroys error data pointlessly.

alice-i-cecile commented on crates/bevy_internal/src/default_plugins.rs:
```diff
@@ -7,6 +7,7 @@ plugin_group! {
         #[cfg(feature = "bevy_log")]
         bevy_log:::LogPlugin,
         bevy_app:::TaskPoolPlugin,
+        bevy_async:::AsyncPlugin,
```
Substantive: I want to discuss how to best expose this functionality to users in a way that can be readily disabled at compile time. This is probably missing a feature flag.

alice-i-cecile commented on crates/bevy_async/src/plugin.rs:
```diff
@@ -0,0 +1,123 @@
+/// driven from a known `SyncPoint` on the world-owning thread.
+///
+/// This supports arbitrary async runtimes as well as multiple Bevy Worlds / Bevy Apps.
+pub struct AsyncPlugin {
```
Substantive: there's no alerting for "user forgot to add a `async_world_sync_point`" anywhere, and there's none added by default. My gut feeling says we should add one by default with this plugin to `First` or something.

alice-i-cecile commented on crates/bevy_async/src/bridge_request.rs:
```diff
@@ -0,0 +1,204 @@
+use bevy_platform::sync::Arc;
+
+/// Drives the queued bridge work for `SyncPoint`.
+///
```
Nit: some missing doc links here.

andriyDev replied:
Done.

alice-i-cecile commented on examples/async_tasks/async_bridge_primitive.rs:
```diff
@@ -0,0 +1,117 @@
+}
+
+/// Rotates the point light around the origin (0, 0, 0)
+fn rotate_light(mut query: Query<&mut Transform, With<PointLight>>, time: Res<Time>) {
```
Cleanup: this distracts from the actual point of the example. Cut please.

andriyDev replied:
Removed rotate_light from async_bridge_primitive.rs.

MalekiRe commented on examples/async_tasks/async_bridge_primitive.rs:
```diff
@@ -0,0 +1,117 @@
+            // Spawn a task on the async compute pool
+            let task = task.clone();
+            pool.spawn(async move {
+                let delay = std::time::Duration::from_secs_f32(rand::rng().random_range(2.0..8.0));
```
This is pulled straight from `async_channel_pattern`/`async_compute`, should those examples also be changed? The idea being to draw a parallel between the two examples and how they can be used.

cart committed:
Merge branch 'main' into malek/async_ecs

andriyDev committed:
Fix nits for the async_bridge_primitive example.

andriyDev commented on crates/bevy_async/src/wake_signal.rs:
```diff
@@ -0,0 +1,63 @@
+        }
+        #[cfg(not(feature = "std"))]
+        {
+            // No-op on std, since we are only using local futures we should tick them
```
What is the intent of this comment @alice-i-cecile? Do you want this cfg block to be deleted? Should we keep this comment (in the code) anyway?

andriyDev replied:
I SPOTTED THE DIFFERENCE. std vs no_std. Got it now lol

andriyDev replied:
Done.

andriyDev committed:
Remove the max_async_ticks_per_sync_point.

andriyDev committed:
Fixup bridge_future docs.

andriyDev committed:
Improve docs on bridge_request.

andriyDev committed:
Add a test to show that async access won't block the ECS **after** the access.

andriyDev committed:
Add a test to show that different sync points will progress different tasks.

andriyDev committed:
Add a test to show that we can spawn many more tasks than threads.

andriyDev commented on crates/bevy_async/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+//!   8. Schedule resumes (normal systems run)
+//!
+//!
+//! Dual locking:
```
IMO there's way too many implementation details that really don't matter to a user of the crate? Like I don't think this dual locking is something that matters to a user at all? We don't make these locks public at all.

I think we can rewrite this to be simpler and more usage-focused.

andriyDev committed:
Rewrite the docs for bevy_async to be usage-focused.

andriyDev committed:
Fix comment referencing std behavior in a no_std block!

andriyDev committed:
Code-format `AsyncBridge` in the example description.

andriyDev commented on Cargo.toml:
```diff
@@ -2211,6 +2211,18 @@ description = "How to use `AsyncComputeTaskPool` to complete longer running task
+name = "async_bridge"
+path = "examples/async_tasks/async_bridge_primitive.rs"
+doc-scrape-examples = true
+wasm = false
```
Looks like this wasm key is unused, which causes a warning whenever running anything on the workspace.

andriyDev committed:
Remove unused wasm key on [[example]].

andriyDev committed:
Revive the "protocol" doc comment as a big block comment inside of async_world_sync_point.

MalekiRe committed:
Merge pull request #5 from andriyDev/async_ecs_nits

Apply comment suggestions on the Async ECS PR.

andriyDev commented on crates/bevy_async/src/bridge_request.rs:
````diff
@@ -0,0 +1,204 @@
+/// ```
+/// The second reason is spoken of prior. Poll may fail to finish for a variety of reasons and
+/// should be given several chances before giving up.
+pub fn async_world_sync_point<SyncPoint: 'static>(world: &mut World) {
````
Done. Also added a test.

andriyDev commented on crates/bevy_async/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+}
+
+#[cfg(test)]
+mod tests {
```
Testing this is actually quite difficult. Single-threaded bevy_tasks will actually ignore the tick budget and will keep running until they hit a non-bridge await that returns Poll::Pending. Multi-threaded also races for access, which may not be reliable.

I have a proposal to simplify this and just entirely delete this max_async_ticks stuff https://github.com/MalekiRe/bevy_malek_async/pull/2. Though maybe we'd still want to bring it back and only process that many ECS accesses per frame? idk.

andriyDev commented on examples/async_tasks/async_bridge_primitive.rs:
```diff
@@ -0,0 +1,117 @@
+//! to offload computationally intensive tasks to a background thread pool and process them
+//! asynchronously.
+//!
+//! Unlike the channel-based approach (where tasks send results directly via a communication
```
Done.

andriyDev commented on examples/async_tasks/async_bridge_primitive.rs:
```diff
@@ -0,0 +1,117 @@
+}
+
+// Number of cubes to spawn across the x, y, and z axis
+const NUM_CUBES: i32 = 6;
```
Rearranged these. NUM_CUBES is now shared between two systems, so it needs to be here.

andriyDev commented on crates/bevy_async/src/bridge_request.rs:
```diff
@@ -0,0 +1,204 @@
+use bevy_ecs::schedule::InternedSystemSet;
+use bevy_platform::sync::Arc;
+
+/// Drives the queued bridge work for `SyncPoint`.
```
Done.

andriyDev commented on crates/bevy_async/src/bridge_request.rs:
````diff
@@ -0,0 +1,204 @@
+///     commands.trigger(PlayerDoesAttack);
+/// }).await;
+/// ```
+/// The second reason is spoken of prior. Poll may fail to finish for a variety of reasons and
````
Done.

andriyDev commented on Cargo.toml:
```diff
@@ -2137,6 +2137,18 @@ description = "How to use `AsyncComputeTaskPool` to complete longer running task
+
+[package.metadata.example.async_bridge]
+name = "Async Bridge"
+description = "An example showing how to offload work to background async tasks using the AsyncBridge primitive."
```
Done.

andriyDev commented on crates/bevy_async/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+//! - The main Bevy schedule
+//! - Futures and async tasks running on other threads
+//!
+//! This is done through the bridge primitive introduced in this crate
```
"The bridge primitive" isn't one thing, it's basically the whole crate. Nevertheless I've rewritten this.

andriyDev commented on crates/bevy_async/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+//! The objective here is to coordinate two participants that want to share World access:
```
Done. I focused more on the usage and what feature this crate gives users.

andriyDev commented on crates/bevy_async/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+//! - If the world is dropped, futures don't leak and eventually finish (in an error state)
+//!
+//!
+//! The protocol:
```
Done.

andriyDev commented on crates/bevy_async/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+//! This is done through the bridge primitive introduced in this crate
+//!
+//!
+//! Invariants of this crate:
```
I just removed it entirely. These are kinda just implementation details.

andriyDev commented on crates/bevy_async/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+//!
+//! Dual locking:
+//!
+//! The published World pointer lock is managed by the `ScopedStatic` primitive in `scoped_static_storage` (only one future can lock this at a time)
```
I just removed this entirely. I wouldn't even call it concurrent, tasks are currently entirely exclusive because they lock the world pointer.

andriyDev commented on crates/bevy_async/src/plugin.rs:
```diff
@@ -0,0 +1,123 @@
+    }
+}
+
+/// Internal resource to manage a limit on how many times we try to drive the async <-> ecs bridge
```
Done.

andriyDev commented on crates/bevy_async/src/plugin.rs:
```diff
@@ -0,0 +1,123 @@
+}
+
+/// Internal resource to manage a limit on how many times we try to drive the async <-> ecs bridge
+/// per sync point.
```
Entirely removed from the AsyncPlugin. I also moved the docs to this resource.

andriyDev commented on crates/bevy_async/src/plugin.rs:
```diff
@@ -0,0 +1,123 @@
+    ///
+    /// We may need to do this multiple times because one task's progress can
+    /// unblock another task that previously returned `Poll::Pending`.
+    pub max_async_ticks_per_sync_point: usize,
```
Removed from the plugin and documented to insert the resource.

andriyDev commented on crates/bevy_async/src/bridge_future.rs:
```diff
@@ -0,0 +1,233 @@
+    /// We bridge *at* the `_sync_point` `SyncPoint` with our `bridge_fn`.
+    pub async fn bridge<BridgeFn, Out, SyncPoint: 'static>(
+        &self,
+        _sync_point: SyncPoint,
```
Done. I prefer the drop so we don't accidentally allow other unused variables.

andriyDev commented on crates/bevy_async/src/bridge_future.rs:
```diff
@@ -0,0 +1,233 @@
+impl<P: SystemParam + 'static> AsyncSystemState<P> {
+    /// Create a new `AsyncSystemState` from an `AsyncWorld` matching the Api surface of
+    /// `SystemState` with `World`.
+    pub fn new(world: AsyncWorld) -> Self {
```
Done.

andriyDev commented on examples/async_tasks/async_bridge_primitive.rs:
```diff
@@ -0,0 +1,117 @@
+    let pool = bevy::tasks::AsyncComputeTaskPool::get();
+
+    // Reuse tasks so you don't have to pay the system init cost every time it runs.
+    let task = async_world.system_state::<(
```
Added a much more descriptive comment here, explaining why we don't need to cache it (we only run this system once!).

andriyDev commented on examples/async_tasks/async_bridge_primitive.rs:
```diff
@@ -0,0 +1,117 @@
+///
+/// The task is offloaded to the `AsyncComputeTaskPool`, allowing heavy computation
+/// to be handled asynchronously, without blocking the main game thread.
+fn setup(
```
Done.

andriyDev committed:
Require futures to queue every time they run.

This is approximately equivalent to https://github.com/MalekiRe/bevy_malek_async/pull/2.

andriyDev committed:
Allow system params to be deduced, and add support for exclusive system params.

This is equivalent to https://github.com/MalekiRe/bevy_malek_async/commit/ceabfba64067fb2772b74a706c4d14861dbb8c33

Co-authored-by: MalekiRe <MalekiRe@users.noreply.github.com>

andriyDev committed:
Fix bridge futures spuriously polled resulting in a panic.

This corresponds to https://github.com/MalekiRe/bevy_malek_async/pull/3

andriyDev committed:
Merge 'main' branch into 'sync-async-ecs'.

andriyDev committed:
Bump bevy_async to 0.20.0-dev.

andriyDev committed:
Fix README examples.

andriyDev committed:
Simplify the doc comments for async_world_sync_point now that we don't have ticks.

andriyDev committed:
Fix doc-comment mentioning a symbol without a doc-link.

andriyDev committed:
Fix lint-allows having no reason, and switch to all_tuples!.

andriyDev committed:
Fix tests and examples to move the system state type to the usage.

andriyDev committed:
Remove stale doc comment.

andriyDev committed:
Remove AssetPlugin::default(), now that it's a unit struct.

andriyDev committed:
Fix tests in bevy_async with multi_threading.

andriyDev committed:
Fix test lint.

andriyDev committed:
Fix dropped_world test having a race condition.

MalekiRe committed:
Merge pull request #6 from andriyDev/sync-async-ecs

Sync async ecs

andriyDev committed:
Write a release notes.

MalekiRe committed:
Merge pull request #8 from andriyDev/async-rel

Add release notes for bevy_async.

cart commented on crates/bevy_async/Cargo.toml:
```diff
@@ -0,0 +1,44 @@
+bevy_tasks = { path = "../bevy_tasks", version = "0.19.0-dev", default-features = false, optional = true }
+
+# other
+scoped_static_storage = { version = "0.1.2", default-features = false }
```
I think I will insist that we include these in `bevy_async` rather than linking out to shared crates. The code is tiny, this is a supply chain risk, and it makes it harder to maintain and evolve the code.
